wavesrv/pkg/blockstore/blockstore.go

```diff
@@ -0,0 +1,449 @@
+}
+
+var cache map[string]*CacheEntry = make(map[string]*CacheEntry)
+var globalLock *sync.Mutex = &sync.Mutex{}
+
+func InsertFileIntoDB(ctx context.Context, fileInfo FileInfo) error {
+	metaJson, err := json.Marshal(fileInfo.Meta)
+	if err != nil {
+		return fmt.Errorf("Error writing file %s to db: %v", fileInfo.Name, err)
+	}
+	txErr := WithTx(ctx, func(tx *TxWrap) error {
+		query := `INSERT INTO block_file VALUES (?, ?, ?, ?, ?, ?, ?, ?)`
+		tx.Exec(query, fileInfo.BlockId, fileInfo.Name, fileInfo.Opts.MaxSize, fileInfo.Opts.Circular, fileInfo.Size, fileInfo.CreatedTs, fileInfo.ModTs, metaJson)
+		return nil
+	})
+	if txErr != nil {
+		return fmt.Errorf("Error writing file %s to db: %v", fileInfo.Name, txErr)
+	}
+	return nil
+}
+
+func WriteFileToDB(ctx context.Context, fileInfo FileInfo) error {
+	metaJson, err := json.Marshal(fileInfo.Meta)
+	if err != nil {
+		return fmt.Errorf("Error writing file %s to db: %v", fileInfo.Name, err)
+	}
+	txErr := WithTx(ctx, func(tx *TxWrap) error {
+		query := `UPDATE block_file SET blockid = ?, name = ?, maxsize = ?, circular = ?, size = ?, createdts = ?, modts = ?, meta = ? where blockid = ? and name = ?`
+		tx.Exec(query, fileInfo.BlockId, fileInfo.Name, fileInfo.Opts.MaxSize, fileInfo.Opts.Circular, fileInfo.Size, fileInfo.CreatedTs, fileInfo.ModTs, metaJson, fileInfo.BlockId, fileInfo.Name)
+		return nil
+	})
+	if txErr != nil {
+		return fmt.Errorf("Error writing file %s to db: %v", fileInfo.Name, txErr)
+	}
+	return nil
+
+}
+
+func WriteDataBlockToDB(ctx context.Context, blockId string, name string, index int, data []byte) error {
+	txErr := WithTx(ctx, func(tx *TxWrap) error {
+		query := `REPLACE INTO block_data values (?, ?, ?, ?) ` //ON CONFLICT (blockid, name) DO UPDATE SET blockid = ?, name = ?, partidx = ?, data = ?`
+		tx.Exec(query, blockId, name, index, data)
+		return nil
+	})
+	if txErr != nil {
+		return fmt.Errorf("Error writing data block to db: %v", txErr)
+	}
+	return nil
+}
+
+func MakeFile(ctx context.Context, blockId string, name string, meta FileMeta, opts FileOptsType) error {
+	curTs := time.Now().UnixMilli()
+	fileInfo := FileInfo{BlockId: blockId, Name: name, Size: 0, CreatedTs: curTs, ModTs: curTs, Opts: opts, Meta: meta}
+	err := InsertFileIntoDB(ctx, fileInfo)
+	if err != nil {
+		return err
+	}
+	curCacheEntry := MakeCacheEntry(&fileInfo)
+	SetCacheEntry(ctx, GetCacheId(blockId, name), curCacheEntry)
+	return nil
+}
+
+func WriteToCacheBlock(ctx context.Context, blockId string, name string, block *CacheBlock, p []byte, pos int, length int, cacheNum int) (int64, int, error) {
+	cacheEntry, found := GetCacheEntry(ctx, blockId, name)
+	if !found {
+		return 0, 0, fmt.Errorf("Cache entry not in cache")
+	}
+	cacheEntry.Lock.Lock()
+	defer cacheEntry.Lock.Unlock()
+	var bytesWritten = 0
+	blockLen := len(block.data)
+	fileMaxSize := cacheEntry.Info.Opts.MaxSize
+	maxWriteSize := fileMaxSize - (int64(cacheNum) * MaxBlockSize)
+	numLeftPad := int64(0)
+	if pos > blockLen {
```

Comment:
So in the case where the offset is greater than the file size, like if we call WriteAt(100) on a new file, the block store will left pad 0's. This creates some options and some potential concerns. 

First of all, do you think I should left pad 0's, or should I just return an error if the write at offset is greater than the file size? 

Currently, what I do is I calculate the left pad based on when I am writing to a cache block. So I only left pad in the current cache rather than the whole file. I add the amount of left pad bytes to the bytesWritten return value. This could lead to some ambiguity in the bytesWritten return value

I have one test where I create a circular file whose max size is 2 blocks
I write an 11 byte message to an offset 4 bytes from the end of the file. 
the code first writes 4 bytes at the end of the second cache, and it left pads enough 0s to fill the entire cache block except for the last 4 bytes. 
Then, the code writes the last 7 bytes of the message to the front of the first cache block. There is no need to left pad any bytes to the cache, so the first cache block ends up being only 7 bytes long, while the second cache block is BlockSize bytes long. The bytesWritten value returned by WriteAt is MaxBlockSize + 7, although the size of the file is now 2 * MaxBlockSize, but the actual stored size of the first cache is only 7 bytes

When the user calls read with offset 0, reading from the beginning of the file, the returned bytesRead value will be only 7 bytes, but it would normally read 0s until the end of the buffer if the code had filled the first cache block with 0s 

The other thing that I could do is left pad the entire file with 0s up to the write position, rather than just the cache, so now there is no potential for ambiguous behavior, but do we think that would be a significant performance hit? 

Say I want to write at offset 5gb in an empty file, I could be writing 5gb of left pad 0s, where if I only left pad the current cache the max left padding that I would be doing is BlockSize

Also, as a secondary general question, should I return the amount of left pad bytes in the return value of WriteAt? I'm thinking maybe it makes sense to just return the actual bytes written so that there is no ambiguity in return value 

Overall, my thoughts on this are that I think we should return an error if offset is greater than file size unless we have a good use case for it, and if so, I think we should keep the cache optimization and also not return the left pad bytes in the return value of WriteAt

Reply:
Update on this - I went with the cache block only left pad approach, and I made it so left pad bytes are not returned in the return value of the WriteAt call

There is ambiguity in the ReadAt return value still. If you read from a cache block that hasn't been left padded yet, but is not the last block in the file, and you try to read some bytes that "should" be 0 but are currently empty, read at doesn't return those bytes 

Psuedocode Example - lets say MaxBlockSize is 5
WriteAt(dataToWrite: "test", offset: 6)
After the write, the cache looks like: 
Cache Entry File Size: 10 
Block1 { data: [], size: 0}
Block2 {data: [0 t e s t, size: 5}

readBuf = make([]byte, 5) 
bytesRead = ReadAt(buf: readBuf, offset: 0)
bytesRead is 0 
readBuf is unchanged
really readBuf should be filled with 0s and bytesRead should be 5 

Should I make a special case for this?

Reply:
Currently I have no special case for this 

this is probably the only "open" for me, unless any reviewers have any issues 

everything else I would say I'm pretty comfortable with